### PR TITLE
Check types of operands of *_bounds_cast operators (#286)

### PR DESCRIPTION
**Pull request by jijoongmoon**
. Add check whether E1 is a pointer type or and integral type. If E1 is
   functional type, it checks the return type of E1.

 . Add parsing bounds expression for bounds cast expression.
   - Add check whether the second argument of bounds cast expression is bounds
     expression such as count(10) or bounds(a, b) and it parses them except
     byte_count expression. If it is byte_count then it return ExprError().

### CODE REVIEW

**Comment by jijoongmoon** [comment]
I think it is ready to review. Thanks!

File: lib/Sema/SemaExpr.cpp [hunk 1]
```diff
@@ -12466,6 +12466,63 @@ ExprResult Sema::CreatePositionalParameterExpr(unsigned Index, QualType QT) {
   return new (Context) PositionalParameterExpr(Index, QT);
 }
 
+bool Sema::CheckBoundsCastBaseType(Expr *E1) {
+  bool Result = false;
+  QualType SrcTy = E1->getType();
+  if (SrcTy->isFunctionType())
```
Comment (dtarditi) [inline lib/Sema/SemaExpr.cpp]: I don't understand why we are checking the return type of E1 if E1 is a function type.    I think we  do not want to allow assume_bounds_cast or  dynamic_bounds cast to or from function types.  I think this line and the following line can be deleted.

File: lib/Sema/SemaExpr.cpp [hunk 2]
```diff
@@ -12526,6 +12586,9 @@ ExprResult Sema::ActOnBoundsCastExprCount(
   ExprResult ResE2 = CorrectDelayedTyposInExpr(E2);
   SourceLocation TypeLoc = (castTInfo->getTypeLoc()).getBeginLoc();
 
+  if(CheckBoundsCastBaseType(E1))
```
Comment (dtarditi) [inline lib/Sema/SemaExpr.cpp]: There needs to be a space after the `if '

File: lib/Parse/ParseExpr.cpp [hunk 3]
```diff
@@ -3097,11 +3097,23 @@ ExprResult Parser::ParseBoundsCastExpression() {
...
     ConsumeToken();
-    syntax = BoundsCastExpr::SyntaxType::Count;
-    E2 = Actions.CorrectDelayedTyposInExpr(ParseAssignmentExpression());
-    if (E2.isInvalid()) {
-      SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
-      return ExprError();
+    if (StartsBoundsExpression(Tok)) {
+      IdentifierInfo *Ident = Tok.getIdentifierInfo();
+
+      if (Ident == Ident_byte_count)
```
Comment (jijoongmoon) [inline lib/Parse/ParseExpr.cpp]: I see. I will delete it.

File: lib/Sema/SemaExpr.cpp (hunk 1 above)
Comment (jijoongmoon) [inline lib/Sema/SemaExpr.cpp]: Yes. I will delete it.

File: lib/Sema/SemaExpr.cpp [hunk 4]
```diff
@@ -12479,6 +12536,9 @@ ExprResult Sema::ActOnBoundsCastExprSingle(
   QualType DestTy = GetTypeFromParser(D, &castTInfo);
   SourceLocation TypeLoc = (castTInfo->getTypeLoc()).getBeginLoc();
 
+  if(CheckBoundsCastBaseType(E1))
```
Comment (jijoongmoon) [inline lib/Sema/SemaExpr.cpp]: I will put space after the 'if'

File: lib/Sema/SemaExpr.cpp (hunk 2 above)
Comment (jijoongmoon) [inline lib/Sema/SemaExpr.cpp]: I will put space after the 'if'

**Comment by jijoongmoon** [comment]
I think it is ready for further review. Thanks!

**Comment by dtarditi** [comment]
Looks good!

**Pull request merged**